rules/Php83/Rector/ClassConst/AddTypeToConstRector.php

```diff
@@ -0,0 +1,237 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\Php83\Rector\ClassConst;
+
+use PhpParser\Node;
+use PhpParser\Node\Name\FullyQualified;
+use PhpParser\Node\Stmt\Class_;
+use PhpParser\Node\Stmt\Interface_;
+use PhpParser\Node\Stmt\Trait_;
+use PHPStan\Reflection\ClassReflection;
+use PHPStan\Reflection\MissingConstantFromReflectionException;
+use PHPStan\Reflection\ReflectionProvider;
+use Rector\Core\Exception\FullyQualifiedNameNotAutoloadedException;
+use Rector\Core\Rector\AbstractRector;
+use Rector\Core\ValueObject\PhpVersionFeature;
+use Rector\VersionBonding\Contract\MinPhpVersionInterface;
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\CodeSample;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+
+/**
+ * @see \Rector\Tests\Php83\Rector\ClassConst\AddTypeToConstRector\AddTypeToConstRectorTest
+ */
+class AddTypeToConstRector extends AbstractRector implements MinPhpVersionInterface
+{
+    public function __construct(
+        private readonly ReflectionProvider $reflectionProvider,
+    ) {
+    }
+
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition('Add const to type', [
+            new CodeSample(
+                <<<'CODE_SAMPLE'
+class SomeClass
+{
+    public const TYPE = 'some_type';
+}
+CODE_SAMPLE
+                ,
+                <<<'CODE_SAMPLE'
+class SomeClass
+{
+    public const string TYPE = 'some_type';
+}
+CODE_SAMPLE
+                ,
+            ),
+        ]);
+    }
+
+    public function getNodeTypes(): array
+    {
+        return [Class_::class, Interface_::class, Trait_::class];
+    }
+
+    /**
+     * @param Class_|Interface_|Trait_ $node
+     */
+    public function refactor(Node $node): Class_|Interface_|Trait_|null
```

Comment:
I am thinking that abstract class, interface, and trait need to be skipped, the guard should be similar to how property guard works

https://github.com/rectorphp/rector-src/blob/765b63696eb3bc240e92f10b00169b146c99463a/rules/TypeDeclaration/Guard/PropertyTypeOverrideGuard.php#L12

The change should be mostly on final class first, then only private const, then protected without override.

You can look how `TypedPropertyFromAssignsRector` works

https://github.com/rectorphp/rector-src/blob/765b63696eb3bc240e92f10b00169b146c99463a/rules/TypeDeclaration/Rector/Property/TypedPropertyFromAssignsRector.php#L35

Reply:
Is there a reason you feel we should be conservative about public constants and not alter Interface/Traits etc? I'm just curious if you have examples of situations where it would be a problem.

Reply:
Here the example https://3v4l.org/1fZSG#v8.3.0 
 which interface has child already typed, but different type, change interface constant will cause fatal error

https://3v4l.org/1fZSG#v8.3.0

Reply:
Thanks @samsonasik . I've made the changes to only work with final or private properties. I'm not sure how I'd make it work with protected properties to detect that it's been overridden?